version that inherits largely from scipy.stats

eacharles opened the pull request:
Please have a look at the demo notebook to see what I have in mind here.

eacharles commented:
Some comments are about where this is:

- I'm pretty happy with the interface between Ensemble the underlying scipy.stats classes.   I think that is pretty close to being locked down.
- I think we probably want to fix up the Pdf_gen interface a bit, and consider moving the stuff like plot_native and convert inside the interface instead of having it be external.
- We should do some performance testing for the classes that derive from scipy.stats and consider if this is the best way to implement this.   
- Need unit tests.
- I think we want to consider how this will be used on large data sets and come up with some lazy evaluation classes that allow you to operate on large datasets that are on disk.

eacharles commented:
Ok, some comments on this PR.

- I've update the setup.py file with the current set of dependencies.
- I've added unit tests and gotten it to 100% coverage, I'm not trying to test every error condition, those are protected with the '#pragma : no cover' marker
  - A few of the unit test that test the self-consistency of the distributions (i.e,. do the pdf match the cdf) could use some tuning
  - A few of the tests that test the accuracy of converting distributions from one type to another could use some tuning.
- I've added a .pylintrc file and gotten in very close to 10/10.   

I think there are a number of open questions:

- Are we happy with the overall structure?
- Are we happy with the names of the classes and functions? If not, now is the time to change them.
- Are we happy with the API and how the main actions are done, i.e., construction, conversion, plotting, writing and reading.
- Are we happy with the extendability?
   - Is it clear how to add a new class?  
   - Is it easy enough to add a new class that most collaborators could do it? 
- We should discuss the consistency and conversion tests.
- How are we going to use this in pz-rail, and elsewhere?

Gschwend commented:
I'd suggest to add a comment about the dependency on LaTeX to run the demo notebook, maybe in the README file, in a markdown cell inside the notebook, or as a warning message when running setup.py.  

I found typos in the markdown cells of demo notebook, they are: 

In section "Extending scipy.stats to implement ...":
 -  "The next part of this **notebokk** shows (...)" 

In sections "qp histogram distribution", "qp interpolated", and "qp spline distribution", the same typos were found:  
- "This represents a set of distributions made by **interoplating** a set of histograms (...) To **construce** this you (...) Note that **the the** native visual **represenation** is different. 

In section "Converting the ensemble": 
- The qp.qp_convert function lets you convert ensembles to other representations. **Do do** this (...) 

In section "Storing and retreiving ensembles": 
- Finally, we can compare the **meoments** of each approximation and compare those to the moments **ofthe** true distribution.

eacharles commented:
Ok, some more comments on my part.  

I think that some of the distribution classes should be simplified.   In particular, the KDE doesn't really store the distribution as a KDE, but rather converts it to a spline, so I think it would be better to provide functions to do that converstion.   
Likewise the flex basis does the decomposition to basis functions, I think that should also be external to the construction.   The representation classes should be simple, and just take data members as input.   Computing those data members should be external, and part of the conversion mechanism.

eacharles commented:
Next comment, we should be clear as to if the conversion acts on an rv_frozen object, or on a Ensemble.   I've find of floated between the two.  Clarity would be good.

eacharles commented:
Hi Sam, sorry, what line were you referring to?   Also, I played around with the implemetation of the functions and found that it is fine to use scipy.interpolate.inter1d for the interpolation both the the case that we have a shared set of X values and different X values for each PDF.   I've pushed those changes to this branch.

sschmidt23 commented:
Oh, sorry, I added a single comment on a specific line thinking it would link to that line.  It is [line 576 in pdf_gen.py](https://github.com/LSSTDESC/qp/blob/211a916f10dc3f0333a81531f326c94dac982313/qp/pdf_gen.py#L576)

sschmidt23 commented:
Ah, on the previous comment, my error was caused because my zgrid had a shape (301, 1), if I flatten the array that is input into interp_fixed so that it has shape (301,) then everything works fine, so this is not actually a problem, sorry for the confusion.

sschmidt23 commented on tests/test_ensemble.py:
```diff
@@ -0,0 +1,101 @@
+        
+        median = ens.median()
+        mean = ens.mean()
+        var = ens.var()
```
var() and std() throw errors for older versions of scipy (1.3 and 1.4), so we should remember to put a requirement on scipy>=1.5.2 in the setup.py file

eacharles replied:
Done

sschmidt23 commented on qp/interp_pdf.py:
```diff
@@ -0,0 +1,267 @@
+interp_fixed = interp_fixed_grid_rows_gen.create
+
+
+class interp_rows_gen(Pdf_rows_gen):
```
Oh, maybe interp_irregular?

eacharles replied:
That sounds good.  interp, and inter_irregular.

sschmidt23 commented on qp/plotting.py:
```diff
@@ -0,0 +1,253 @@
+STYLES['quantiles'] = '--'#(0,(5,10))
+STYLES['histogram'] = ':'#(0,(3,6))
+STYLES['samples'] = '-.'#(0,(1,2))
+
```
Add a 'flex' color/style?

eacharles replied:
Good idea, I'll add that to the branch with flex.

sschmidt23 replied:
resolving since flex is no longer covered in this PR

sschmidt23 reviewed (changes requested):
Very good stuff, there is one change where we need to add a np.expand_dims so that a test doesn't fail, but otherwise everything is working.  A couple additional comments:

In the demo notebook, maybe we should explicitly mention any conversions that are known to be highly inefficient/slow so users know that they may not want to use those on large scale datasets.

Currently the flex ensemble is commented out in demo.ipynb, if it is to be uncommented and included, then I would explicitly put in an nbasis=~9, and maybe add a caveat in a comment about potential overfitting if nbasis is much greater than the number of evaluation points (Note: this may be moot if the sparse deconvolution method replaces the current least squares fit and that method does not suffer from the same overfitting).

I created some demo PDFs by running RAIL/estimation/flexcode on the small tomo_challenge files.  I think something similar with some more complex, non-Gaussian, distributions might make for a good demonstration of some features.  Also, when computing CDFs, PIT values, and such for a few hundred of these PDFs and comparing the distributions after converting the ensembles, I see some non-insignificant differences in the CDF and PIT values.  Again, this might be a nice demo to illustrate potential limitations effects of grid spacing, number of basis functions, etc... in an expanded demo notebook.

The switch to separate files for each distribution is much more readable.  I will leave comments on naming mostly to Alex, though I am wondering if "_rows" needs to be included in the names?

eacharles commented on qp/flex_pdf.py:
```diff
@@ -0,0 +1,289 @@
+            basis = evaluate_basis(np.expand_dims(x_trans, -1), self.coefs.shape[1], self.basis_system)
+            return np.matmul(self._coefs[rr], basis.T).flatten()
+        x_trans = box_transform(xr, self.a, self.b)
+        basis = evaluate_basis(x_trans, self.coefs.shape[1], self.basis_system)
```
Thanks for all the work you did with the flex basis stuff.  I pulled the flex out of this PR at Alex's request to cut out that dependency, but I've saved the work into a new branch u/eacharles/eac-dev-v3, which I'll push to git shortly.

eacharles commented:
1. I’ve cleaned up the name space and the factory methods.  Now there is a single factory that manages all the distribution classes, and the static calls like qp.read, qp.create, qp.convert.   Also, I simplified the class names.
2. I did some meta coding and wrote a piece of code to pull in and wrap all of the scipy.stats distributions and insert them into the factory.   The wrapped distributions are accessible vi e.g., qp.stats.norm or qp.stats.lognorm  See the demo notebook for examples. 
3. I did some more meta coding and used the factory to auto generate tests for any class that has a test_data attribute attached to it. 
4. I added documentation
5. I updated the notebook to use the new names

sschmidt23 commented:
Updates look good, there is one mistake in demo.ipynb in the cell demonstrating the qp histogram distribution (cell 15, I believe): a remaining instance of `hist_pdf` should be changed to `hist_dist`.

In demo.ipynb you have the example for qp.interp, do you think that we need an example for qp.interp_irregular as well?  qp.interp_irregular is slightly less intuitive, but I also don't see nearly as many potential use cases for it, so I'm torn as to whether it needs to be included in the demo.

I am about to commit a couple of superficial typo fixes to contributing.rst, other than the typos it looks good.  I did notice while looking through in the Step 3 example: `check_input` seems to be for checking whether distributions are normalized for hist and interp, that's not the most intuitive name, maybe `check_input_norm`?



-

aimalz commented:
Sorry I'm still working through this review! I have a comment that isn't tied to any specific line of code, namely some lingering confusion over the parameters for the interpolated, KDE, and spline parameterizations. I think this might be another instance of me getting hung up on some notion of the physical meaning that's uncorrelated with the logical way to code them up, but, in case it matters, I keep thinking that the keyword arguments/parameters of the spline parameterization, for example, should be the properties of the knots, rather than the (x, y) values that constitute the grid evaluations used to produce them, meaning a spline-parameterized PDF object would be produced by the conversion of a PDF object defined by grid evaluations to the spline parameterization. That being said, I think even if a change like this were made, it wouldn't have a big impact on the overall API and structure, so perhaps doesn't need to be addressed for this PR?

aimalz commented on qp/conversion.py:
```diff
@@ -0,0 +1,151 @@
+
+CONVERSIONS = ConversionDict()
+
+def qp_convert(obj_from, class_to, method=None, **kwargs):
```
Were we going to rename this to just `convert`?

eacharles replied:
That's been removed.

eacharles commented on qp/plotting.py:
```diff
@@ -0,0 +1,253 @@
+
+
+
+def plot_pdf_quantiles_on_axes(axes, xvals, yvals, quantiles, **kwargs):
```
It is nice to have all dependencies on matplotlib in a single file, as that avoids gotcha's with initializing it in several places.   I was thinking to have all the stuff that actually depends on matplotlib live in the plotting.py as utility functions and then just have the various representations call to the functions that they need to implement plot() and plot_native()

aimalz replied:
That sounds like a good solution.

aimalz replied:
Oh, also, the quantile parameterization is supposed to be plotted as a linear interpolation -- sorry for not noting that before!

eacharles replied:
Hm..   Can you write down the equation that you have in mind to define the PDF at a given point in terms of the quantile values?   I.e., given a set of quantiles you can define the CDF at particular points, but if you assume the CDF is smoothly interpolated between the quantiles, then the PDF looks like a series of step functions.   If you want to the PDF to be interpolated, then you have to make some assumption about how to go from the quantiles to the CDF/ PDF.

eacharles commented on qp/hist_pdf.py:
```diff
@@ -0,0 +1,154 @@
+        """
+        axes, _, kw = get_axes_and_xlims(**kwargs)
+        vals = pdf.dist.pdfs[pdf.kwds['row']]
+        return plot_pdf_histogram_on_axes(axes, hist=(pdf.dist.bins, vals), **kw)
```
See above for why I was thinking to have all the plotting functions in one place.   But yes, we should give the histogram plot to better represent a histogram.

eacharles replied:
Ok, I've updated the histogram plotting representation.

eacharles commented on qp/utils.py:
```diff
@@ -1,44 +1,16 @@
-import scipy as sp
+
 from scipy import stats as sps
+from scipy.interpolate import splev, splrep, interp1d
```
I think this depends a bit on if they might be reused elsewhere.   For the splines I don't think that is the case, and we should move them, but for the interpolation and histogram normalization I'm not so sure.

eacharles replied:
So, I've put a couple of histogram evaluation functions in utils as well ( they are useful also for _pdf evaluation of the quantile rep).

eacharles replied:
Ok, I've moved the spline-related functions to the spline_pdf.py file.

eacharles commented on setup.py:
```diff
@@ -1,8 +1,10 @@
     name="qp",
-    version="0.1",
+    version=qp.__version__,
     author="Alex Malz, Phil Marshall",
```
Done

eacharles replied:
I've added my name.

aimalz commented on docs/notebooks/demo.ipynb:
```diff
@@ -238,39 +350,64 @@
+    "print(\"The input and output shapes are:\", xvals.shape, kde_vals.shape)\n",
+    "# Construct a single PDF for plotting\n",
+    "kde_dist1 = qp.spline_from_samples(xvals=xvals_kde, samples=np.atleast_2d(samples[0]))\n",
+    "fig, axes = qp.plotting.plot_native(kde_dist1, xlim=(-5., 5.), label=\"kde\")\n",
```
I get a type error here `only integer scalar arrays can be converted to a scalar index` that seems to be caused by scipy not getting what it's expecting in a `pdf.pdf()` call in `plotting.py`.

eacharles replied:
It works for me.  What version of scipy are you using?

eacharles commented on qp/pdf_gen.py:
```diff
@@ -0,0 +1,402 @@
+        return plot_dist_pdf(pdf, **kwargs)
+
+    @classmethod
+    def plot_native(cls, pdf, **kwargs):
```
Has this been fixed?

aimalz reviewed (approved):
At this point, the big changes are solid! Side issues that came up along the way (quantile parameterization matching that of the old qp, enhancing built-in plots, adding back FlexCode support, file i/o approach, updating citation, annotating demo, etc.) are separable from the big API and back-end changes so can be addressed in subsequent PRs. Bravo @eacharles!

sschmidt23 reviewed (approved):
The couple of things I commented on earlier are minor questions on notation and examples, so I say we put them off so that we can get this large PR merged.  S
